server/pairing/challenge_management.go

```diff
@@ -0,0 +1,247 @@
+		return nil, err
+	}
+
+	return &ChallengeGiver{
+		cookieStore: cs,
+		encryptor:   e.Renew(),
+		logger:      logger,
+	}, nil
+}
+
+func (cg *ChallengeGiver) getIP(r *http.Request) (net.IP, *ChallengeError) {
+	h, _, err := net.SplitHostPort(r.RemoteAddr)
+	if err != nil {
+		cg.logger.Error("getIP: h, _, err := net.SplitHostPort(r.RemoteAddr)", zap.Error(err), zap.String("r.RemoteAddr", r.RemoteAddr))
+		return nil, &ChallengeError{"error", http.StatusInternalServerError}
+	}
+	return net.ParseIP(h), nil
+}
+
+func (cg *ChallengeGiver) registerClientIP(r *http.Request) *ChallengeError {
+	hIP, err := cg.getIP(r)
+	if err != nil {
+		return err
+	}
+	cg.authedIP = hIP
+	return nil
+}
+
+func (cg *ChallengeGiver) validateClientIP(r *http.Request) *ChallengeError {
+	// If we haven't registered yet register the IP
+	if cg.authedIP == nil || len(cg.authedIP) == 0 {
+		err := cg.registerClientIP(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Then compare the current req RemoteIP with the authed IP
+	hIP, err := cg.getIP(r)
+	if err != nil {
+		return err
+	}
+
+	if !cg.authedIP.Equal(hIP) {
+		cg.logger.Error(
+			"request RemoteAddr does not match authedIP: expected '%s', received '%s'",
+			zap.String("expected", cg.authedIP.String()),
+			zap.String("received", hIP.String()),
+		)
+		return &ChallengeError{"forbidden", http.StatusForbidden}
+	}
+	return nil
+}
+
+func (cg *ChallengeGiver) getSession(r *http.Request) (*sessions.Session, *ChallengeError) {
+	s, err := cg.cookieStore.Get(r, sessionChallenge)
+	if err != nil {
+		cg.logger.Error("checkChallengeResponse: cg.cookieStore.Get(r, sessionChallenge)", zap.Error(err), zap.String("sessionChallenge", sessionChallenge))
+		return nil, &ChallengeError{"error", http.StatusInternalServerError}
+	}
+	return s, nil
+}
+
+func (cg *ChallengeGiver) generateNewChallenge(s *sessions.Session, w http.ResponseWriter, r *http.Request) ([]byte, *ChallengeError) {
+	challenge := make([]byte, 64)
+	_, err := rand.Read(challenge)
+	if err != nil {
+		cg.logger.Error("regenerateNewChallenge: _, err = rand.Read(challenge)", zap.Error(err))
+		return nil, &ChallengeError{"error", http.StatusInternalServerError}
+	}
+
+	s.Values[sessionChallenge] = challenge
+	err = s.Save(r, w)
+	if err != nil {
+		cg.logger.Error("regenerateNewChallenge: err = s.Save(r, w)", zap.Error(err))
+		return nil, &ChallengeError{"error", http.StatusInternalServerError}
+	}
+
+	return challenge, nil
+}
+
+func (cg *ChallengeGiver) block(s *sessions.Session, w http.ResponseWriter, r *http.Request) *ChallengeError {
+	s.Values[sessionBlocked] = true
+	err := s.Save(r, w)
+	if err != nil {
+		cg.logger.Error("block: err = s.Save(r, w)", zap.Error(err))
+		return &ChallengeError{"error", http.StatusInternalServerError}
+	}
+
+	return &ChallengeError{"forbidden", http.StatusForbidden}
+}
+
+func (cg *ChallengeGiver) checkChallengeResponse(w http.ResponseWriter, r *http.Request) *ChallengeError {
+	ce := cg.validateClientIP(r)
```

Comment:
I do try to follow the guidelines 😆 , I regularly sight them in my own reviews of other PRs.

`cErr` seems better, the only reason that I don't use the `err` var here is because `cErr` is an explicit `error` type `ChallengeError` and this value conflicts with function calls that do return only the untyped `error` interface.

The reason I don't return the `error` interface from these functions is because the consuming code uses the properties of `ChallengeError` to create `http.Error()`s .

However this seems like a place where a change to the whole strategy makes sense. I'll still return `ChallengeError`s within the function, but all functions will return the `error` interface. Consuming code will type check the error and if it is a `ChallengeError` process a `http.Error()`, if not throw a `http.StatusInternalServerError`.

Reply:
`c` means "challenge"